### DISCUSSION TITLE
Memory CPU spikes

### DISCUSSION THREAD

**Discussion by musabmasood** [argoproj/argo-cd#6964] [Q&A]
Our production ArgoCD instance connects to 3 external k8s clusters. It manages around 90 `Applications`, which are a mix of helm and kustomize.

This is 1 month's activity dashboard in Datadog. I notice memory and CPU spikes in the Application Controller and repo server. I couldn't find any other relation than the fact that it happens around syncs (obvious). Any recommendations on smoothing out the spikes?

We are on ArgoCD `v2.0.3`.

![image](https://user-images.githubusercontent.com/8170066/129117894-d3e20078-5a02-4d8f-8b57-c98d61e7ec0a.png)

**Comment by jgwest** [comment]
Hi @musabmasood, I've not seen any specific guidance around reducing memory spikes, but generally they would be related to:
- Cluster resources: Retrieving large number of cluster resources and storing them in various caches within the argo cd processes (for example, as part of a new cluster, or a large number of changes to an existing cluster).
- Sync operations: Comparing the live state with the on-cluster target state, for a large number of application resources (eg a large number of new applications being deployed/modified/etc)

Many of the related tunables for Argo CD processes can be found [here](https://argoproj.github.io/argo-cd/operator-manual/high_availability/#argocd-application-controller).

It will take some experimentation on your part to see if these work for you, but the parameters that jump out at me from that page are `--status-processors` and `--operation-processors` on the Application Controller. These parameters are responsible for the number of concurrent goroutines that are handling the above 2 operations. 

So you may consider reducing those, as my guess is that doing so will spread out the memory usage over a longer period of time. BUT, you should know that this will (by design) slow down the time it takes for Argo CD to complete a sync operation.

**Reply by jgwest** [reply]
And one more suggestion: `--kubectl-parallelism-limit` may be also be used to limit the number of concurrent resource application processes, which can tend to be large especially with large manifests.

**Reply by alexmt** [reply]
App syncing causes big memory spike (and probably CPU spike as well) in application controller. This had been improved in upcoming v2.1: controller no longer need any additional memory to sync resources and we pretty much can deprecate `--kubectl-parallelism-limit` flag.

**Reply by jgwest** [reply]
Most of the above answer was related to the Application Controller spike. And my colleague @jannfis pointed out [this issue](https://github.com/argoproj/argo-cd/issues/5505) where this was previously discussed.

For reposerver there is `--parallelismlimit` to limit the number of concurrent manifest tool invocations, but (for example) if you are using a monorepo with a large number of applications, some of this might be unavoidable due to the sheer volume of resources that need to be generated by the tool (see monorepo scaling concerns).

**Reply by musabmasood** [reply]
> Most of the above answer was related to the Application Controller spike. And my colleague @jannfis pointed out [this issue](https://github.com/argoproj/argo-cd/issues/5505) where this was previously discussed.
> 
> For reposerver there is `--parallelismlimit` to limit the number of concurrent manifest tool invocations, but (for example) if you are using a monorepo with a large number of applications, some of this might be unavoidable due to the sheer volume of resources that need to be generated by the tool (see monorepo scaling concerns).

We are using the `argocd.argoproj.io/manifest-generate-paths` annotation aggressively to avoid that, so that it doesn't generate every single manifest on every commit.

@alexmt thanks for this, keenly waiting for the `2.1` release to come out 🎉
